storageccl: fix a correctness bug in Import

danhhz opened the pull request:
Previously, Import would construct WriteBatches by iterating through
each file in the import and adding keys to the batch. Then it would
periodically send a WriteBatch request when the batch size reached a
threshold. WriteBatch clears any existing data in the affected keyrange
when applied, so later files could overwrite data in earlier ones. This
creates incorrect data at the end of the import.

MultiIterator multiplexes iteration over the sstables in parallel and
returns the latest version of each key (skipping that key if the latest
version is a tombstone). This means each keyrange is only affected by
WriteBatch once.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14203)
<!-- Reviewable:end -->

danhhz committed:
engine: fix Iterator.{UnsafeKey,UnsafeValue} comments

danhhz committed:
engineccl: add an iterator that multiplexes engine.Iterators

danhhz committed:
storageccl: use MultiIterator in Import

Previously, Import would construct WriteBatches by iterating through
each file in the import and adding keys to the batch. Then it would
periodically send a WriteBatch request when the batch size reached a
threshold. WriteBatch clears any existing data in the affected keyrange
when applied, so later files could overwrite data in earlier ones. This
creates incorrect data at the end of the import.

MultiIterator multiplexes iteration over the sstables in parallel and
returns the latest version of each key (skipping that key if the latest
version is a tombstone). This means each keyrange is only affected by
WriteBatch once.

Closes #14152.

danhhz commented:
@dt I'd still like to expand the tests for Import, but wanted to get this out for early feedback. The first two commits are ready for review (and you can see how they'll be used in the third.)

dt commented:
Reviewed 5 of 5 files at r1.
Review status: 5 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 37 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNE5XL0LcrGe0oanHA:-KfNE5XL0LcrGe0oanHB:b12a8dh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L37)):*
> ```Go
> // MakeMultiIterator creates an iterator that multiplexes engine.Iterators.
> func MakeMultiIterator(iters []engine.Iterator) *MultiIterator {
> 	// TODO(dan): Support Next().
> ```

nit: This TODO seems out of place.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 97 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNUQDZTEaKVakA6phg:-KfNUQDZTEaKVakA6phh:bcy6976) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L97)):*
> ```Go
> 			key := iter.UnsafeKey()
> 			cmp := bytes.Compare(key.Key, f.key.Key)
> 			if cmp < 0 {
> ```

this method might be a case where I'd get value from narrative commenting.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 100 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNUyDlYElav_y0wjyp:-KfNUyDlYElav_y0wjyq:bkypjng) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L100)):*
> ```Go
> 				f.itersToAdvance = f.itersToAdvance[:0]
> 				f.itersToAdvance = append(f.itersToAdvance, iterIdx)
> 				f.key.Key = append(f.key.Key[:0], key.Key...)
> ```

would it make sense just to keep the idx of the iter that currently looks like the winner and only do the work of copying the key and value once we've finished checking them all?

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 114 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNVbNjnQ8q4waFoehx:-KfNVbNjnQ8q4waFoehy:b-hbrjpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L114)):*
> ```Go
> 						"got two entries for the same key and timestamp %s: %x vs %x",
> 						f.key, f.value, iter.UnsafeValue(),
> 					)
> ```

return?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r1.
Review status: 5 of 8 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 23 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNa3O1HwSNbceEqrjF:-KfNa3O25Wk8xJR7Jo9j:b1lia64) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L23)):*
> ```Go
> type MultiIterator struct {
> 	iters          []engine.Iterator
> 	itersToAdvance []int
> ```

this deserves a comment explaining that it's only here to save allocations.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNaBe-kP_zkLu8Di38:-KfNaBe-kP_zkLu8Di39:b-f12f5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L25)):*
> ```Go
> 	itersToAdvance []int
> 
> 	valid bool
> ```

these fields all duplicate a similar field from the underlying iterators, which feels a little dirty.

Perhaps it'd be cleaner to use references to the underlying iterator from which these values are drawn. In the case of `valid`, I think it's always true so long as any underlying iterator is true?

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 31 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNYNuB-r7d9YP1e7OT:-KfNYNuB-r7d9YP1e7OU:b-na5xoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L31)):*
> ```Go
> }
> 
> // TODO(dan): This is intentionally a subset of the engine.Iterator interface.
> ```

would be useful to name the methods omitted.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 97 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNUQDZTEaKVakA6phg:-KfNahtmaYhP-72WUwyU:bi6n0b5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L97)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

this method might be a case where I'd get value from narrative commenting.
</blockquote></details>

agreed.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 100 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNUyDlYElav_y0wjyp:-KfNaj_bOVS5qU1H1UgJ:b-1bk4ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L100)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

would it make sense just to keep the idx of the iter that currently looks like the winner and only do the work of copying the key and value once we've finished checking them all?
</blockquote></details>

heh, made the same suggestion in person.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203)*
<!-- Sent from Reviewable.io -->

dt commented on pkg/ccl/storageccl/import.go:
```diff
@@ -117,49 +118,51 @@ func evalImport(ctx context.Context, cArgs storage.CommandArgs) error {
+		if log.V(3) {
+			log.Infof(ctx, "Put %s -> %s", key.Key, value.PrettyPrint())
 		}
-		if err := iter.Error(); err != nil {
```
We want to keep this, no?

danhhz commented:
Thanks for the reviews! This is much more readable now. RFAL

I updated the ImportTest. I feel like there's more work that could be done there, but this is pretty good for now.

> We want to keep this, no?
Indeed. Done.

---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 23 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNa3O1HwSNbceEqrjF:-KfStKyqOIoSared-AZW:b-wvcxyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L23)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this deserves a comment explaining that it's only here to save allocations.
</blockquote></details>

Now saves state between calls

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNaBe-kP_zkLu8Di38:-KfStQfRIkOlTEC_De1A:b-g36pcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L25)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

these fields all duplicate a similar field from the underlying iterators, which feels a little dirty.

Perhaps it'd be cleaner to use references to the underlying iterator from which these values are drawn. In the case of `valid`, I think it's always true so long as any underlying iterator is true?
</blockquote></details>

Agreed, but implementing Valid by calling it on all the underlying iterators seems wasteful, so I'm going to leave it like this. We can't get around the err field, anyway.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 31 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNYNuB-r7d9YP1e7OT:-KfSti6TZG-87kMByWmM:b-clxy64) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L31)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

would be useful to name the methods omitted.
</blockquote></details>

It's possible that you got something different from this than I intended, so reworked it a bit. Also, not sure I like how easily that list could rot.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 37 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNE5XL0LcrGe0oanHA:-KfStvHME-ktFVjsF-bB:bw0webt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L37)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: This TODO seems out of place.
</blockquote></details>

Did it while this was all loaded up in my head. We'll need it for point in time restores.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 97 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNUQDZTEaKVakA6phg:-KfSu-6nDVeexphwrrHi:b-pass81) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L97)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

agreed.
</blockquote></details>

Done. Happy to polish it further if anything is still unclear.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 100 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNUyDlYElav_y0wjyp:-KfSu4U_M76Eq10gLQ3x:by4mggw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L100)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

heh, made the same suggestion in person.
</blockquote></details>

Done. This made it pretty easy to implement Next and NextKey. Good job gents!

I also changed NextKey such that it doesn't skip over tombstones (and updated the use in the third commit). The new behavior matches what other iterator implementations do.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 114 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNVbNjnQ8q4waFoehx:-KfSuRigAkg37hkN7FDD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L114)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

return?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-:-KfSup2OeQyEiI6IQPsB:b-vegwuk)*
<!-- Sent from Reviewable.io -->

tamird commented on pkg/storage/engine/engine.go:
```diff
@@ -67,11 +67,11 @@ type Iterator interface {
 	// ValueProto unmarshals the value the iterator is currently
```
heh, this commit just updates comments now :)

danhhz commented:
@dt ping on this? Any further production testing of backup/restore is effectively blocked on getting this in

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

the `:= func(` test helpers still scare me with their ability to shadow/close over state, but 

---

Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator_test.go, line 24 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgF8D70YIDjwBeiWMG:-KfgF8D70YIDjwBeiWMH:b8ucztc) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator_test.go#L24)):*
> ```Go
> 
> func TestMultiIterator(t *testing.T) {
> 	rocksDB := engine.NewInMem(roachpb.Attributes{}, 1<<20)
> ```

no leaktest?

---

*[pkg/ccl/storageccl/engineccl/multi_iterator_test.go, line 94 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgEZeTps1foYkHWsRd:-KfgEZeTps1foYkHWsRe:b-mjjfq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator_test.go#L94)):*
> ```Go
> 			}
> 
> 			check := func(expected string, fn func(*MultiIterator)) func(*testing.T) {
> ```

nit: i'm still slightly -1 on inline funcs that can shadow / close over state but ¯\_(ツ)_/¯

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-:-KfgDyfE9J6QrFl1W2aY:b16eh18)*
<!-- Sent from Reviewable.io -->

dt commented:
Whoops, ignore that comment here, i moved it inline.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-:-KfgFI4qcK0sZWNyubAG:b-tsfitb)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 6 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNaBe-kP_zkLu8Di38:-KfgDERLOsfWwErFt6un:b-z6c1x5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L25)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Agreed, but implementing Valid by calling it on all the underlying iterators seems wasteful, so I'm going to leave it like this. We can't get around the err field, anyway.
</blockquote></details>

Wouldn't you evict iterators as they become invalid? In that way, if any iterators remain in`iters`, then the iterator is valid?

In any case, both these fields need comments explaining their behaviour.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 70 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgE7FsoETOQwZpqoW-:-KfgE7FsoETOQwZpqoW0:bv14h8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L70)):*
> ```Go
> }
> 
> // UnsafeKey returns the same value as Key, but the memory is invalidated on the
> ```

Key() isn't a thing.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 76 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgE9eyihslPdMD18c8:-KfgE9eyihslPdMD18c9:bn137tu) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L76)):*
> ```Go
> }
> 
> // UnsafeValue returns the same value as Value, but the memory is invalidated on
> ```

Value isn't a thing.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgIMxo0hLuYWsgVy7X:-KfgIMxo0hLuYWsgVy7Y:bfm9qv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L109)):*
> ```Go
> 	// underlying iterators are exhausted and so is this one.
> 	proposedNextIdx := -1
> 	for iterIdx := range f.iters {
> ```

small nit: the `iterIdx, iter` form might be useful here

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 110 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgITSazPqEFsp65GfR:-KfgITSazPqEFsp65GfS:b-wj1kqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L110)):*
> ```Go
> 	proposedNextIdx := -1
> 	for iterIdx := range f.iters {
> 		// If this iterator is exhausted skip it (or error if it's errored).
> ```

if it's exhausted you could evict it, no?

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 146 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgJDR6wlJTyurRE9lk:-KfgJDR6wlJTyurRE9ll:b-a8rxtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L146)):*
> ```Go
> 			} else if proposedMVCCKey.Timestamp == iterMVCCKey.Timestamp {
> 				// We have two exactly equal mvcc keys (both key and timestamps
> 				// match). It's ambiguous which should sort first, so error.
> ```

you could avoid this error by having currentIdx be a slice, right? As long as the values are the same, there's no error, right?

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgJO34pq0m8PpxvPgD:-KfgJO34pq0m8PpxvPgE:bnol7b7) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L147)):*
> ```Go
> 				// We have two exactly equal mvcc keys (both key and timestamps
> 				// match). It's ambiguous which should sort first, so error.
> 				f.valid = false
> ```

no need for this, right? setting the error is enough.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator_test.go, line 94 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgEZeTps1foYkHWsRd:-KfgJc6di7W0C-k5MiZh:b-c98my5) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator_test.go#L94)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: i'm still slightly -1 on inline funcs that can shadow / close over state but ¯\_(ツ)_/¯
</blockquote></details>

+1, would prefer a freestanding function.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator_test.go, line 117 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgJmPjuRyoD8EHH4oP:-KfgJmPjuRyoD8EHH4oQ:b-wguv46) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator_test.go#L117)):*
> ```Go
> 				}
> 			}
> 			t.Run("NextKey", check(test.expectedNextKey, func(it *MultiIterator) {
> ```

fun fact: this `func(it ...) {}` can be written `(*MultiIterator).NextKey`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Moved `slurpSSTablesLatestKey` out too, since it was trivial. Left `writeSST` since passing it a *testing.T or returning an error were both gross (and it doesn't close over much)

Thanks for the reviews!

---

Review status: 1 of 5 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfNaBe-kP_zkLu8Di38:-KfgfIaQhgXc2x7v8AYR:b-wf1o0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3111fb3dab16f7479b7caf7bb52220eb99972b5f/pkg/ccl/storageccl/engineccl/multi_iterator.go#L25)):*
> Wouldn't you evict iterators as they become invalid? In that way, if any iterators remain initers, then the iterator is valid?

Ah, makes sense. Left this one as a TODO (down in the advance impl)

> In any case, both these fields need comments explaining their behaviour.

Done

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 70 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgE7FsoETOQwZpqoW-:-KfgbWr3J3Az0fk3ICG5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Key() isn't a thing.
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 76 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgE9eyihslPdMD18c8:-KfgbWeg9p--bBUYCo06:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L76)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Value isn't a thing.
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgIMxo0hLuYWsgVy7X:-Kfga3rifBhLGee6eOIE:bngs2l9) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L109)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

small nit: the `iterIdx, iter` form might be useful here
</blockquote></details>

I had that originally, but thought the parallelism here was nice. Switched back.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 110 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgITSazPqEFsp65GfR:-KfgoD7tn9oQ6X02qvKt:b-6osuet) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L110)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

if it's exhausted you could evict it, no?
</blockquote></details>

Left this one as a TODO

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 146 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgJDR6wlJTyurRE9lk:-Kfg_x6F0sMK-yWYxJlk:b-el7vn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L146)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you could avoid this error by having currentIdx be a slice, right? As long as the values are the same, there's no error, right?
</blockquote></details>

Seems like a lot of complexity for a use case we don't need yet. I'd rather this be loud for now

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgJO34pq0m8PpxvPgD:-Kfg_wYyCH8QUj2jQDqV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L147)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

no need for this, right? setting the error is enough.
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator_test.go, line 24 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgF8D70YIDjwBeiWMG:-KfgYhwCOnIXFmzGBZ9P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator_test.go#L24)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

no leaktest?
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator_test.go, line 94 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgEZeTps1foYkHWsRd:-Kfg_ZoqpIKtqlzIwMQ4:bvzjjvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator_test.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

+1, would prefer a freestanding function.
</blockquote></details>

I played with this for a few minutes but couldn't find a version I liked as a freestanding function. Made it a second level of table based test instead.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator_test.go, line 117 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgJmPjuRyoD8EHH4oP:-KfgYaq0KcXXERjsxZjc:bb741sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator_test.go#L117)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

fun fact: this `func(it ...) {}` can be written `(*MultiIterator).NextKey`.
</blockquote></details>

TIL!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-:-Kfgma9fAAzEyfm0hhBV:b-wnsl1k)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r4, 2 of 2 files at r5.
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgIMxo0hLuYWsgVy7X:-KfgpKG55BUfF0zP78px:b-oys26e) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L109)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I had that originally, but thought the parallelism here was nice. Switched back.
</blockquote></details>

What parallelism?

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 28 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-Kfgpb9up50p_2C3yNRY:-Kfgpb9up50p_2C3yNRZ:b23p8b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7957d61dbe1e1d6ea91e59c9b6db10cbb7e92500/pkg/ccl/storageccl/engineccl/multi_iterator.go#L28)):*
> ```Go
> 
> 	// valid is true if currentIdx points to something meaningful.
> 	valid bool
> ```

heh, it now seems that `valid` is exactly equivalent to `currentIdx > -1`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-KfgIMxo0hLuYWsgVy7X:-KfgqipSfUHRz-uY6kpt:bph6x78) ([raw file](https://github.com/cockroachdb/cockroach/blob/809e483d178744f511a0aaae2bd06b4ec71302cd/pkg/ccl/storageccl/engineccl/multi_iterator.go#L109)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What parallelism?
</blockquote></details>

Everything was accessed via `f.iters[someIdx]`

---

*[pkg/ccl/storageccl/engineccl/multi_iterator.go, line 28 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203#-Kfgpb9up50p_2C3yNRY:-Kfh6Tpy2bmM4H3WSlzX:bj57faw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7957d61dbe1e1d6ea91e59c9b6db10cbb7e92500/pkg/ccl/storageccl/engineccl/multi_iterator.go#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

heh, it now seems that `valid` is exactly equivalent to `currentIdx > -1`
</blockquote></details>

Good point. removed valid

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14203)*
<!-- Sent from Reviewable.io -->

danhhz commented:
TFTRs! This looks much better than when I started
